Error trying to compile 0.77.1 release under Haiku

### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### Different version than latest?

_No response_

### What Operating System are you using?

Other

### If Other OS, please describe

Haiku

### Relevant hardware info

_No response_

### Have you checked that no other similar issue already exists?

- [X] I have searched and not found similar issues.

### A clear and concise description of what the bug is.

Hi!  When trying to compile Dosbox-staging r. 0.77.1, I got the following error:


```
src/hardware/libhardware.a(serialport_misc_util.cpp.o): In function TCPClientSocket::TCPClientSocket(int)':
/boot/home/Compilar/dosbox-staging-0.77.1/build/../src/hardware/serialport/misc_util.cpp:50: undefined reference to `getpeername'
/boot/home/Compilar/dosbox-staging-0.77.1/build/../src/hardware/serialport/misc_util.cpp:59: undefined reference to `getsockname'
collect2: error: ld returned 1 exit status
[228/229] Compiling C++ object tests/support.p/support_tests.cpp.o
ninja: build stopped: subcommand failed.


```

Thanks for your time!


### Steps to reproduce the behaviour.

Explain how to reproduce
1. extract the 0.77.1 source code .zip file.
2. run "meson setup -Duse_opengl=false build" from Terminal
3. run "ninja -C build" from Terminal


### Your configuration

_No response_

### Provide a Log

```text
[227/229] Linking target dosbox
FAILED: dosbox 
c++  -o dosbox dosbox.p/meson-generated_.._version.cpp.o dosbox.p/src_main.cpp.o dosbox.p/src_dosbox.cpp.o -Wl,--as-needed -Wl,--no-undefined '-Wl,-rpath,$ORIGIN/subprojects/munt-libmt32emu_2_5_0:/packages/libsdl2-2.0.14-6/.self/develop/lib:/packages/opusfile-0.12-1/.self/develop/lib:/packages/fluidsynth2-2.1.8-1/.self/develop/lib' -Wl,-rpath-link,/boot/home/Compilar/dosbox-staging-0.77.1/build/subprojects/munt-libmt32emu_2_5_0 -Wl,-rpath-link,/packages/libsdl2-2.0.14-6/.self/develop/lib -Wl,-rpath-link,/packages/opusfile-0.12-1/.self/develop/lib -Wl,-rpath-link,/packages/fluidsynth2-2.1.8-1/.self/develop/lib -Wl,--start-group src/cpu/libcpu.a src/debug/libdebug.a src/dos/libdos.a src/libs/decoders/libdecoders.a src/fpu/libfpu.a src/gui/libgui.a src/libs/ppscale/libppscale.a src/hardware/libhardware.a src/libs/nuked/libnuked.a src/libs/residfp/libresidfp.a src/ints/libints.a src/midi/libmidi.a subprojects/munt-libmt32emu_2_5_0/libmt32emu.so src/misc/libmisc.a src/shell/libshell.a -latomic /packages/libsdl2-2.0.14-6/.self/develop/lib/libSDL2.so /packages/opusfile-0.12-1/.self/develop/lib/libopusfile.so /packages/libpng16-1.6.37-1/.self/develop/lib/libpng16.so /packages/libpng16-1.6.37-1/.self/develop/lib/libz.so /packages/sdl2_net-2.0.1-2/.self/develop/lib/libSDL2_net.so /packages/sdl2_net-2.0.1-2/.self/develop/lib/libSDL2.so /packages/fluidsynth2-2.1.8-1/.self/develop/lib/libfluidsynth.so -Wl,--end-group
src/hardware/libhardware.a(serialport_misc_util.cpp.o): In function `TCPClientSocket::TCPClientSocket(int)':
/boot/home/Compilar/dosbox-staging-0.77.1/build/../src/hardware/serialport/misc_util.cpp:50: undefined reference to `getpeername'
/boot/home/Compilar/dosbox-staging-0.77.1/build/../src/hardware/serialport/misc_util.cpp:59: undefined reference to `getsockname'
collect2: error: ld returned 1 exit status
[228/229] Compiling C++ object tests/support.p/support_tests.cpp.o
ninja: build stopped: subcommand failed.
```


### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

Hi, thanks for your answer!  I tried again with the main branch, but this time I got another error:

`
[45/260] Compiling C++ object src/libs/loguru/libloguru.a.p/loguru.cpp.o
FAILED: src/libs/loguru/libloguru.a.p/loguru.cpp.o 
c++ -Isrc/libs/loguru/libloguru.a.p -Isrc/libs/loguru -I../src/libs/loguru -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wnon-virtual-dtor -Wextra -Wpedantic -std=c++17 -g -Wmaybe-uninitialized -Weffc++ -Wextra-semi -MD -MQ src/libs/loguru/libloguru.a.p/loguru.cpp.o -MF src/libs/loguru/libloguru.a.p/loguru.cpp.o.d -o src/libs/loguru/libloguru.a.p/loguru.cpp.o -c ../src/libs/loguru/loguru.cpp
../src/libs/loguru/loguru.cpp: In function 'loguru::Text loguru::vtextprintf(const char*, va_list)':
../src/libs/loguru/loguru.cpp:431:30: error: 'vasprintf' was not declared in this scope; did you mean 'vsprintf'?
  431 |                 int result = vasprintf(&buff, format, vlist);
      |                              ^~~~~~~~~
      |                              vsprintf
../src/libs/loguru/loguru.cpp: In function 'loguru::Text loguru::errno_as_text()':
../src/libs/loguru/loguru.cpp:572:22: warning: unused variable 'buff' [-Wunused-variable]
  572 |                 char buff[256];
      |                      ^~~~
../src/libs/loguru/loguru.cpp: In function 'void loguru::init(int&, char**, const loguru::Options&)':
../src/libs/loguru/loguru.cpp:622:38: warning: unused variable 'this_thread' [-Wunused-variable]
  622 |                                 auto this_thread = pthread_self();
      |                                      ^~~~~~~~~~~
../src/libs/loguru/loguru.cpp: In function 'void loguru::set_thread_name(const char*)':
../src/libs/loguru/loguru.cpp:1035:42: warning: unused parameter 'name' [-Wunused-parameter]
 1035 |         void set_thread_name(const char* name)
      |                              ~~~~~~~~~~~~^~~~
../src/libs/loguru/loguru.cpp: In function 'void loguru::get_thread_name(char*, long long unsigned int, bool)':
../src/libs/loguru/loguru.cpp:1076:25: error: 'pthread_getname_np' was not declared in this scope
 1076 |                         pthread_getname_np(pthread_self(), buffer, length);
      |                         ^~~~~~~~~~~~~~~~~~
../src/libs/loguru/loguru.cpp:1097:66: error: invalid conversion from 'pthread_t' {aka '_pthread_thread*'} to 'uint64_t' {aka 'long long unsigned int'} [-fpermissive]
 1097 |                                 uint64_t thread_id = pthread_self();
      |                                                      ~~~~~~~~~~~~^~
      |                                                                  |
      |                                                                  pthread_t {aka _pthread_thread*}
[50/260] Compiling C++ object src/dos/libdos.a.p/program_rescan.cpp.o
ninja: build stopped: subcommand failed.
`

`
[45/260] Compiling C++ object src/libs/loguru/libloguru.a.p/loguru.cpp.o
FAILED: src/libs/loguru/libloguru.a.p/loguru.cpp.o 
c++ -Isrc/libs/loguru/libloguru.a.p -Isrc/libs/loguru -I../src/libs/loguru -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wnon-virtual-dtor -Wextra -Wpedantic -std=c++17 -g -Wmaybe-uninitialized -Weffc++ -Wextra-semi -MD -MQ src/libs/loguru/libloguru.a.p/loguru.cpp.o -MF src/libs/loguru/libloguru.a.p/loguru.cpp.o.d -o src/libs/loguru/libloguru.a.p/loguru.cpp.o -c ../src/libs/loguru/loguru.cpp
../src/libs/loguru/loguru.cpp: In function 'loguru::Text loguru::vtextprintf(const char*, va_list)':
../src/libs/loguru/loguru.cpp:431:30: error: 'vasprintf' was not declared in this scope; did you mean 'vsprintf'?
  431 |                 int result = vasprintf(&buff, format, vlist);
      |                              ^~~~~~~~~
      |                              vsprintf
../src/libs/loguru/loguru.cpp: In function 'loguru::Text loguru::errno_as_text()':
../src/libs/loguru/loguru.cpp:572:22: warning: unused variable 'buff' [-Wunused-variable]
  572 |                 char buff[256];
      |                      ^~~~
../src/libs/loguru/loguru.cpp: In function 'void loguru::init(int&, char**, const loguru::Options&)':
../src/libs/loguru/loguru.cpp:622:38: warning: unused variable 'this_thread' [-Wunused-variable]
  622 |                                 auto this_thread = pthread_self();
      |                                      ^~~~~~~~~~~
../src/libs/loguru/loguru.cpp: In function 'void loguru::set_thread_name(const char*)':
../src/libs/loguru/loguru.cpp:1035:42: warning: unused parameter 'name' [-Wunused-parameter]
 1035 |         void set_thread_name(const char* name)
      |                              ~~~~~~~~~~~~^~~~
../src/libs/loguru/loguru.cpp: In function 'void loguru::get_thread_name(char*, long long unsigned int, bool)':
../src/libs/loguru/loguru.cpp:1076:25: error: 'pthread_getname_np' was not declared in this scope
 1076 |                         pthread_getname_np(pthread_self(), buffer, length);
      |                         ^~~~~~~~~~~~~~~~~~
../src/libs/loguru/loguru.cpp:1097:66: error: invalid conversion from 'pthread_t' {aka '_pthread_thread*'} to 'uint64_t' {aka 'long long unsigned int'} [-fpermissive]
 1097 |                                 uint64_t thread_id = pthread_self();
      |                                                      ~~~~~~~~~~~~^~
      |                                                                  |
      |                                                                  pthread_t {aka _pthread_thread*}
[50/260] Compiling C++ object src/dos/libdos.a.p/program_rescan.cpp.o
ninja: build stopped: subcommand failed.
`

Hi! About the **loguru**, I followed @LowLevelMahn suggestion and it's correct, i got the same error:

`
```
~/Compilar/loguru-2.1.0/loguru_example> ./build_and_run.sh 
CMake Deprecation Warning at CMakeLists.txt:1 (cmake_minimum_required):
  Compatibility with CMake < 2.8.12 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


-- The C compiler identification is GNU 11.2.0
-- The CXX compiler identification is GNU 11.2.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /boot/system/bin/x86/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /boot/system/bin/x86/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- CMAKE_BUILD_TYPE: Debug
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE  
-- Configuring done
-- Generating done
-- Build files have been written to: /boot/home/Compilar/loguru-2.1.0/loguru_example/build
[ 50%] Building CXX object CMakeFiles/loguru_example.dir/main.cpp.o
In file included from /boot/home/Compilar/loguru-2.1.0/loguru_example/main.cpp:5:
/boot/home/Compilar/loguru-2.1.0/loguru_example/../loguru.cpp: In function 'loguru::Text loguru::vtextprintf(const char*, va_list)':
/boot/home/Compilar/loguru-2.1.0/loguru_example/../loguru.cpp:391:30: error: 'vasprintf' was not declared in this scope; did you mean 'vsprintf'?
  391 |                 int result = vasprintf(&buff, format, vlist);
      |                              ^~~~~~~~~
      |                              vsprintf
/boot/home/Compilar/loguru-2.1.0/loguru_example/../loguru.cpp: In function 'loguru::Text loguru::errno_as_text()':
/boot/home/Compilar/loguru-2.1.0/loguru_example/../loguru.cpp:528:22: error: unused variable 'buff' [-Werror=unused-variable]
  528 |                 char buff[256];
      |                      ^~~~
/boot/home/Compilar/loguru-2.1.0/loguru_example/../loguru.cpp: In function 'void loguru::init(int&, char**, const loguru::Options&)':
/boot/home/Compilar/loguru-2.1.0/loguru_example/../loguru.cpp:578:38: error: unused variable 'this_thread' [-Werror=unused-variable]
  578 |                                 auto this_thread = pthread_self();
      |                                      ^~~~~~~~~~~
/boot/home/Compilar/loguru-2.1.0/loguru_example/../loguru.cpp: In function 'void loguru::set_thread_name(const char*)':
/boot/home/Compilar/loguru-2.1.0/loguru_example/../loguru.cpp:919:42: error: unused parameter 'name' [-Werror=unused-paramete]
  919 |         void set_thread_name(const char* name)
      |                              ~~~~~~~~~~~~^~~~
/boot/home/Compilar/loguru-2.1.0/loguru_example/../loguru.cpp: In function 'void loguru::get_thread_name(char*, long long unsigned int, bool)':
/boot/home/Compilar/loguru-2.1.0/loguru_example/../loguru.cpp:979:54: error: invalid conversion from '_pthread_thread*' to 'uint64_t' {aka 'long long unsigned int'} [-fpermissive]
  979 |                                 uint64_t thread_id = thread;
      |                                                      ^~~~~~
      |                                                      |
      |                                                      _pthread_thread*
cc1plus: all warnings being treated as errors
CMakeFiles/loguru_example.dir/build.make:75: recipe for target 'CMakeFiles/loguru_example.dir/main.cpp.o' failed
make[2]: *** [CMakeFiles/loguru_example.dir/main.cpp.o] Error 1
CMakeFiles/Makefile2:82: recipe for target 'CMakeFiles/loguru_example.dir/all' failed
make[1]: *** [CMakeFiles/loguru_example.dir/all] Error 2
Makefile:90: recipe for target 'all' failed
make: *** [all] Error 2
~/Compilar/loguru-2.1.0/loguru_example> 
```
`

@kcgen : about the **loguru-haiku** branch, I also hit another issue with the **whereami** library:

```
~/Compilar/dosbox-staging-kc-loguru-haiku-1> ninja -C build
ninja: Entering directory `build'
[49/260] Generating symbol file subprojects/munt-libmt32emu_2_5_3/libmt32emu.so.p/libmt32emu.so.symbols
WARNING: Symbol extracting has not been implemented for this platform. Relinking will always happen on source changes.
[94/260] Compiling C object src/libs/whereami/libwhereami.a.p/whereami.c.o
FAILED: src/libs/whereami/libwhereami.a.p/whereami.c.o 
cc -Isrc/libs/whereami/libwhereami.a.p -Isrc/libs/whereami -I../src/libs/whereami -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wextra -Wpedantic -g -MD -MQ src/libs/whereami/libwhereami.a.p/whereami.c.o -MF src/libs/whereami/libwhereami.a.p/whereami.c.o.d -o src/libs/whereami/libwhereami.a.p/whereami.c.o -c ../src/libs/whereami/whereami.c
../src/libs/whereami/whereami.c:790:2: error: #error unsupported platform
  790 | #error unsupported platform
      |  ^~~~~
[99/260] Compiling C++ object src/dos/libdos.a.p/cdrom_image.cpp.o
ninja: build stopped: subcommand failed.
~/Compilar/dosbox-staging-kc-loguru-haiku-1> 
```

I will try to investigate more about this two issues from this libs.  Maybe I could get some help from the Haiku team too.
Thanks a lot for your help!!!

--- a/src/libs/loguru/loguru.cpp
+++ b/src/libs/loguru/loguru.cpp
@@ -420,18 +420,27 @@ namespace loguru
 	LOGURU_PRINTF_LIKE(1, 0)
 	static Text vtextprintf(const char* format, va_list vlist)
 	{
-#ifdef _WIN32
-		int bytes_needed = _vscprintf(format, vlist);
+		// Determine the required length of the string
+		int bytes_needed = 0;
+
+		// Calling vsnprintf is assumed to 'consume' the vlist, therefore we need a copy
+		// of the args to avoid corrupting the stack in the second call to vsnprintf.
+		va_list argcopy;
+		va_copy(argcopy, vlist);
+		bytes_needed = vsnprintf(NULL, 0, format, argcopy);
+		va_end(argcopy);
 		CHECK_F(bytes_needed >= 0, "Bad string format: '%s'", format);
-		char* buff = (char*)malloc(bytes_needed+1);
-		vsnprintf(buff, bytes_needed+1, format, vlist);
+
+		// Allocate the string's buffer
+		++bytes_needed; // Add space for null terminator
+		auto buff = static_cast<char*>(malloc(bytes_needed));
+		CHECK_F(buff != nullptr, "Out of memory");
+
+		// Construct the string and check the result
+		const auto written = vsnprintf(buff, bytes_needed, format, vlist);
+		CHECK_F(written <= bytes_needed, "Bad string format: '%s'", format);
+
 		return Text(buff);
-#else
-		char* buff = nullptr;
-		int result = vasprintf(&buff, format, vlist);
-		CHECK_F(result >= 0, "Bad string format: '" LOGURU_FMT(s) "'", format);
-		return Text(buff);
-#endif
 	}
 
 	Text textprintf(const char* format, ...)
@@ -582,6 +591,7 @@ namespace loguru
 		return Text(STRDUP(buff));
 	#else
 		// Not thread-safe.
+		(void)buff; // unused parameter
 		return Text(STRDUP(strerror(errno)));
 	#endif
 	}
@@ -630,6 +640,9 @@ namespace loguru
 						pthread_set_name_np(this_thread, main_thread_name);
 					#elif defined(__linux__) || defined(__sun)
 						pthread_setname_np(this_thread, main_thread_name);
+					#else
+						// platforms that we don't know how to set the name on
+						(void)this_thread; // unused
 					#endif
 				}
 			#endif // LOGURU_PTHREADS
@@ -1046,6 +1059,9 @@ namespace loguru
 				pthread_set_name_np(pthread_self(), name);
 			#elif defined(__linux__) || defined(__sun)
 				pthread_setname_np(pthread_self(), name);
+			#else
+				// Platforms that may not support setting a thread name
+				(void)name; // unused
 			#endif
 		#elif LOGURU_WINTHREADS
 			// Store thread name in a thread-local storage:
@@ -1073,7 +1089,12 @@ namespace loguru
 			// Ask the OS about the thread name.
 			// This is what we *want* to do on all platforms, but
 			// only some platforms support it (currently).
-			pthread_getname_np(pthread_self(), buffer, length);
+			#if defined(__APPLE__) || defined(__FreeBSD__) || defined(__OpenBSD__) || defined(__linux__) || defined(__sun)
+				pthread_getname_np(pthread_self(), buffer, length);
+			#else
+				// Other platforms that don't support thread names
+				buffer[0] = 0;
+			#endif
 		#elif LOGURU_WINTHREADS
 			snprintf(buffer, static_cast<size_t>(length), "%s", thread_name_buffer());
 		#else
@@ -1094,7 +1115,7 @@ namespace loguru
 				long thread_id;
 				(void)thr_self(&thread_id);
 			#elif LOGURU_PTHREADS
-				uint64_t thread_id = pthread_self();
+				const auto thread_id = reinterpret_cast<uintptr_t>(pthread_self());
 			#else
 				// This ID does not correllate to anything we can get from the OS,
 				// so this is the worst way to get the ID.
@@ -1286,9 +1307,14 @@ namespace loguru
 			}
 		}
 		if (g_preamble_pipe && pos < out_buff_size) {
-			// Because this is the last if-statement, we avoid incrementing the
-			// position to clarify it's unused.
 			(void)snprintf(out_buff + pos, out_buff_size - pos, "| ");
+
+			// Because this is the last if-statement, we avoid incrementing the
+			//  position to clarify it's unused. If more cases are added, then:
+			// int bytes = snprintf(...)
+			// if (bytes > 0) {
+			// 	pos += bytes;
+			// }
 		}
 	}
 
@@ -1367,8 +1393,13 @@ namespace loguru
 			}
 		}
 		if (g_preamble_pipe && pos < out_buff_size) {
-			// Avoid incrementing the position to clarify it's unused
 			(void)snprintf(out_buff + pos, out_buff_size - pos, "| ");
+
+			// Avoid incrementing the position to clarify it's unused
+			// int bytes = snprintf(...)
+			// if (bytes > 0) {
+			// 	pos += bytes;
+			// }
 		}
 	}
 

